keychain_apple.mm

```diff
@@ -63,84 +76,278 @@ static ErrorDescription fromStatus(OSStatus status)
+                                 object:nil];
     }
+    return self;
+}
 
-    if (dataRef)
-        CFRelease(dataRef);
+- (void)dealloc
+{
+    [NSNotificationCenter.defaultCenter removeObserver:self];
+    [super dealloc];
 }
 
-void WritePasswordJobPrivate::scheduledStart()
+- (void)keychainTaskFinished:(NSNotification *)notification
 {
-    NSDictionary *const query = @{
-            (__bridge id) kSecClass: (__bridge id) kSecClassGenericPassword,
-            (__bridge id) kSecAttrService: (__bridge NSString *) service.toCFString(),
-            (__bridge id) kSecAttrAccount: (__bridge NSString *) key.toCFString(),
-    };
+    NSParameterAssert(notification);
+    NSDictionary * const userInfo = notification.userInfo;
+    NSAssert(userInfo, @"Keychain task finished with error notification should contain nonnull user info dictionary");
 
-    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef) query, nil);
+    NSNumber * const retrievedNotificationId = (NSNumber *)[userInfo objectForKey:KeychainNotificationUserInfoNotificationId];
+    if (retrievedNotificationId.unsignedIntValue != _notificationId) {
+        return;
+    }
 
-    if (status == errSecSuccess) {
-        NSDictionary *const update = @{
-                (__bridge id) kSecValueData: (__bridge NSData *) data.toCFData(),
-        };
+    if (_job) {
+        _job->emitFinished();
+    }
 
-        status = SecItemUpdate((__bridge CFDictionaryRef) query, (__bridge CFDictionaryRef) update);
-    } else {
-        NSDictionary *const insert = @{
-                (__bridge id) kSecClass: (__bridge id) kSecClassGenericPassword,
-                (__bridge id) kSecAttrService: (__bridge NSString *) service.toCFString(),
-                (__bridge id) kSecAttrAccount: (__bridge NSString *) key.toCFString(),
-                (__bridge id) kSecValueData: (__bridge NSData *) data.toCFData(),
-        };
+    [self release];
+}
 
-        status = SecItemAdd((__bridge CFDictionaryRef) insert, nil);
+- (void)keychainReadTaskFinished:(NSNotification *)notification
+{
+    NSParameterAssert(notification);
+    NSDictionary * const userInfo = notification.userInfo;
+    NSAssert(userInfo, @"Keychain task finished with error notification should contain nonnull user info dictionary");
+
+    NSNumber * const retrievedNotificationId = (NSNumber *)[userInfo objectForKey:KeychainNotificationUserInfoNotificationId];
+    if (retrievedNotificationId.unsignedIntValue != _notificationId) {
+        return;
+    }
+
+    _privateJob->data.clear();
+    _privateJob->mode = JobPrivate::Binary;
+
+    NSData * const retrievedData = (NSData *)[userInfo objectForKey:KeychainNotificationUserInfoDataKey];
+    if (retrievedData != nil) {
+        if (_privateJob) {
+            _privateJob->data = QByteArray::fromNSData(retrievedData);
+        }
+
+        const CFDataRef dataRef = (__bridge CFDataRef)retrievedData;
+        if (dataRef) {
+            CFRelease(dataRef);
```

Comment:
Here I get a warning from clang analyzer: "Incorrect decrement of the reference count of an object that is not owned at this point by the caller". Is that correct, or bogus?

Reply:
I rewrote the releasing of this now, no more warning :)